### ISSUE TITLE
Need to re-interpret things from run24 analysis notebook

### ISSUE THREAD

**Issue by dwhswenson** [choderalab/octomore#18]
I'm updating the run24 analysis notebook to match new OPS namings (API changes over the last 6 months). It looks like a safety check has been added that we were violating. Specifically, we were analyzing trajectories from Abl as if they were trajectories from Src.

Fortunately, the wrong topology now raises an error (well, if the number of atoms are wrong), so we catch it.

I remember Sonya and I thinking that the values of the DFG angle I was getting seemed weird compared to her histograms, and we ended up with very strange state boundaries compared to those histograms. This would be why.

It looks like there is a transition in this definition of DFG angle (I believe this is the Roux definition, right @sonyahanson?) in clone 3. I'll properly update the notebook in a PR soon (at conference through Thurs), but thought I'd mention the news right away!

**Comment by sonyahanson** [comment]
Ahhhh! Good to know! Yeah, this is still the Roux definition.

**Comment by dwhswenson** [comment]
Quick update (before I go to bed, because I assume you might still be around, Sonya): using clone 3 still only gave partial information. It had one switch, but doesn't tell the full story of "switch-in, switch-out" that we need to calculate an example of the "lifetime" of the state. I've just downloaded no-solvent data from 11401, runs 8 and 19 (following @sonyahanson's lead in #17). Even there, we know that this definition is far from perfect (again, from #17).

My focus now is on finishing the "module" about the `TrajectoryTransitionAnalysis` (formerly called `SingleTrajectoryAnalysis`). Ideally, I'd like to have an example where I can calculate the lifetime of a state (requiring a B-A-B transition). However, I'm starting to think that the Octomore-related data might be added after the official module. I still think this will have some really great data to add as an example for calculating the lifetime, but I think that identifying that data within the next week or two might be difficult.

For the module, I have another example that only focuses on the flux (not the lifetimes/rates). That example is from the Watson-Crick to Hoogsteen base pair transition in DNA. That project is pretty complete -- enough so that I gave a talk on it today at the Dutch national chemistry conference. So we can use that data for an example, even if it only covers the flux calculation.

In summary, if we can find a nice way to get that B-A-B type transition soon, I would love it and would add it as part of the module submission. If we can't find it soon, I think we're find it eventually, and I would still want to add that as a good example. But in that case, time is not pressing.

(Relevant links to [the "module" I'm describing](https://gitlab.e-cam2020.eu/E-CAM/Classical-MD-Modules/merge_requests/4) -- in particular [the file that can be viewed here](https://gitlab.e-cam2020.eu/E-CAM/Classical-MD-Modules/blob/4c5fa67aa5a9239d66cf5c6a3a7fd68a3e8f4f4d/ops_flux_rate_analysis/readme.rst), waiting to fill in examples.)

**Comment by sonyahanson** [comment]
Ahhh, yeah, interesting. I've done a bit more analysis of these trajectories, for example generating images like this that I don't think I've actually pushed anywhere because right now the data is separated by run rather than trajectory:

![abl_11400_inout](https://cloud.githubusercontent.com/assets/6333823/21033816/5fc0e79e-bd81-11e6-8bf1-82d5ff9d96b0.png)

Soon, I'm planning on both (1) cleaning this analysis up and probably putting it into kinalysis (along with some functions that define these coordinates so they can be pulled into things like OPS or MSM building easily) and (2) defining a better distanced-based definition of the DFG flip that we can use for all of this.

Let me know what of this sounds most useful to you.

**Comment by dwhswenson** [comment]
Questions for clarification of that picture: that is, as labeled, 11400? (And the Roux definition?) In that case, it looks like one run has a good B-A-B trajectory (run8? around 50000-60000 on the x-axis) by this definition. (checking that I get the right data before downloading it!) And is the red line your divider between in/out?

It doesn't mean this definition is good (in fact, from the mess in run 25, I don't think that it is), but I might be able to use that data and that definition to show how the code works.

How are the multiple clones handled in that picture? Are they pasted together sequentially? I assume from the name that "clones" all share the same initial configuration, but different velocities/randomization seeds (and are not actually sequential continuations of the previous) -- please correct me if that assumption is wrong!

**Comment by sonyahanson** [comment]
>Questions for clarification of that picture: that is, as labeled, 11400? (And the Roux definition?) In that case, it looks like one run has a good B-A-B trajectory (run8? around 50000-60000 on the x-axis) by this definition. (checking that I get the right data before downloading it!) And is the red line your divider between in/out?

I believe this is all correct!

Yes, clones are pasted together sequentially, but it may be a bit jumbled depending on how it's counting (this is one of the things I don't like about this figure), and yes, clones should be same initial configuration, but different seed.

**Comment by dwhswenson** [comment]
@sonyahanson : I'm a bit confused by the plot that you show above. I downloaded `no-solvent/11400/run8-*`, and get completely different results from what you show. Here are mine (all clones plotting on top of each other:

![image](https://cloud.githubusercontent.com/assets/8178546/21109251/79bc81f2-c098-11e6-8f96-cdf3e3853c32.png)

You can see that nothing goes all the way down to the DFG = -2.0 that you have. I checked that the files were correct (at least, the sizes all matched exactly when I did `ls -lS` on hal and in my local directory, and I even checked MD5 on `clone0`). Are we using the same atoms for the dihedral? I have `[2257,2255,2265,2270]` for Abl.

I got equal nonsense when I ran `run24` as well. Here I only have 20 clones (I had 40 for `run8`). Plotting my results as dots because of periodic boundary hopping:

![image](https://cloud.githubusercontent.com/assets/8178546/21109407/3d0a047c-c099-11e6-9c3d-cf036b3477b8.png)

Looks nothing like yours. Any idea what's going on?

**Comment by sonyahanson** [comment]
Hmmm. Will look into this. Probably will add to current PR.

**Comment by sonyahanson** [comment]
Hi David, it's possible, in fact maybe likely, that what is labeled as 'run8' here is not in fact run8. These plots are made from an array I produced simply to separate runs by whether they started in DFG-in or DFG-out, and I wasn't carefully considering run numbering at the time. The numbering is then based on the index from this array. That's definitely misleading, sorry about that.

Indeed we are defining the dihedrals the same for ABL: https://github.com/choderalab/kinalysis/blob/master/inputs.py#L41

The `DFG_dihedral_byrun` function is in kinalysis here: https://github.com/choderalab/kinalysis/blob/master/kinalysis.py#L35

```python
def DFG_dihedral_byrun(files,def_DFG):

    # Since we are going to sort files by where they are in first frame of clone0
    #   we can only analyze trajectories with a clone0 present.

    path_base = files.split('*')[0]

    clone0_files = "%s/*clone0.h5" % path_base

    globfiles = glob(clone0_files)

    runs_list = []

    for filename in globfiles:
        run_string = re.search('run([^-]+)',filename).group(1)
        run = int(run_string)
        if run not in runs_list:
            runs_list.append(run)
        runs_list.sort()

    dihedral = []
    dihedral_combinetrajs = []

    for run in runs_list:

        trajectories = dataset.MDTrajDataset("%s/run%d-*.h5" % (path_base,run))
        print "Run %s has %s trajectories." % (run,len(trajectories))

        for traj in trajectories:

            dihedral_combinetrajs.append(md.compute_dihedrals(traj,[def_DFG]))
        # flatten
        dihedral_combinetrajs = [val for sublist in dihedral_combinetrajs for val in sublist]

        dihedral.append(dihedral_combinetrajs)
        dihedral_combinetrajs = []

    dihedral = np.asarray([dihedral])

    return [dihedral]
```

Going to work on getting something better out for you here in the next few hours.

**Comment by sonyahanson** [comment]
Indeed, doing this a bit more methodically, I get a run8 that looks much more like yours.

![abl_11400_dfg_by_traj_run8](https://cloud.githubusercontent.com/assets/6333823/21123502/186dedbe-c0a7-11e6-979a-c36d71c87c03.png)

I'll have plots like this for all 100 runs by the morning, as generated by this script: https://github.com/choderalab/octomore/blob/0925883851366035fe0aca0384e205a58b53399b/dihedral_bytraj/DFG_bytraj_allruns.py

**Comment by sonyahanson** [comment]
Run 17 looks promising:

![abl_11400_dfg_by_traj_run17](https://cloud.githubusercontent.com/assets/6333823/21123652/f7c06726-c0a7-11e6-8e8e-84a2c285d747.png)

And indeed my run 24 looks like yours:

![abl_11400_dfg_by_traj_run24](https://cloud.githubusercontent.com/assets/6333823/21123655/02cf146e-c0a8-11e6-92cd-ba941c5a0a07.png)

**Comment by sonyahanson** [comment]
Okay, all these figures are now pushed to this PR: https://github.com/choderalab/octomore/pull/17. This is a bit brute force, but that's all right for now. Other runs that are potentially interesting: 57, 92, 94, and 98.

**Comment by dwhswenson** [comment]
Thanks, @sonyahanson! I'm looking through these to find any examples of B-A-B transitions (and I might fake the definition of the state so that there is one).

However, I think that the real states of interest are centered around -3 (and wrapping around barrier) and centered around +0.5 or +1, right? In that case, one of the clones of run 23 looks very interesting....

There are two things here: (1) finding a B-A-B transition for my example; (2) finding actual flips between DFG in and DFG out. By tweaking the state definition, I can probably find a solution to (1), but the part that's actually scientifically interesting is (2). Right now, I'm focused on the less interesting problem of making a nice example notebook.

**Comment by sonyahanson** [comment]
Agreed that we're working on 1, but it's a good intermediate for now. These plots now make it infinitely easier to find 2. I'm also still interested in moving away from the dihedral, esp if it turns out that we aren't finding 2 by looking at these plots, which is what I'm half expecting. 

Agreed about run 23!
![run23](https://raw.githubusercontent.com/choderalab/octomore/6980c8fb3948c2043d63973cdbed7f0ddc0f5359/dihedral_bytraj/ABL_11400_DFG_by_traj_run23.png)
